daemon/containerd/image_children.go

```diff
@@ -3,216 +3,56 @@ package containerd
 import (
 	"context"
 
-	"github.com/containerd/containerd/content"
-	cerrdefs "github.com/containerd/containerd/errdefs"
 	containerdimages "github.com/containerd/containerd/images"
-	"github.com/containerd/containerd/platforms"
-	"github.com/containerd/log"
 	"github.com/docker/docker/errdefs"
 	"github.com/docker/docker/image"
 	"github.com/opencontainers/go-digest"
-	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
 	"github.com/pkg/errors"
 )
 
-// Children returns a slice of image ID which rootfs is a superset of the
-// rootfs of the given image ID, excluding images with exactly the same rootfs.
+// Children returns a slice of image IDs that are children of the `id` image
 // Called from list.go to filter containers.
 func (i *ImageService) Children(ctx context.Context, id image.ID) ([]image.ID, error) {
-	target, err := i.resolveDescriptor(ctx, id.String())
-	if err != nil {
-		return []image.ID{}, errors.Wrap(err, "failed to get parent image")
-	}
-
-	cs := i.client.ContentStore()
-
-	allPlatforms, err := containerdimages.Platforms(ctx, cs, target)
-	if err != nil {
-		return []image.ID{}, errdefs.System(errors.Wrap(err, "failed to list platforms supported by image"))
-	}
-
-	parentRootFS := []ocispec.RootFS{}
-	for _, platform := range allPlatforms {
-		rootfs, err := platformRootfs(ctx, cs, target, platform)
-		if err != nil {
-			if !cerrdefs.IsNotFound(err) {
-				log.G(ctx).WithFields(log.Fields{
-					"error":    err,
-					"image":    target.Digest,
-					"platform": platform,
-				}).Warning("failed to get platform-specific rootfs")
-			}
-			continue
-		}
-
-		parentRootFS = append(parentRootFS, rootfs)
-	}
-
-	imgs, err := i.client.ImageService().List(ctx)
+	imgs, err := i.client.ImageService().List(ctx, "labels."+imageLabelClassicBuilderParent+"=="+string(id))
 	if err != nil {
 		return []image.ID{}, errdefs.System(errors.Wrap(err, "failed to list all images"))
 	}
 
 	children := []image.ID{}
 	for _, img := range imgs {
-	nextImage:
-		for _, platform := range allPlatforms {
-			rootfs, err := platformRootfs(ctx, cs, img.Target, platform)
-			if err != nil {
-				if !cerrdefs.IsNotFound(err) {
-					log.G(ctx).WithFields(log.Fields{
-						"error":    err,
-						"image":    img.Target.Digest,
-						"platform": platform,
-					}).Warning("failed to get platform-specific rootfs")
-				}
-				continue
-			}
-
-			for _, parentRoot := range parentRootFS {
-				if isRootfsChildOf(rootfs, parentRoot) {
-					children = append(children, image.ID(img.Target.Digest))
-					break nextImage
-				}
-			}
-		}
+		children = append(children, image.ID(img.Target.Digest))
 	}
 
 	return children, nil
 }
```

Comment:
We're changing the contract for this, right? (this is why we're changing the godoc, or just to remove no longer accurate description of the implementation detail?).

Is this a safe change – return all (even indirect) children of this image -> only return direct descendants? Are all the callers expecting this?

If so, then we should also make this "contract" change clearer in the godoc.

(please disregard this comment if I got the changes wrong 😅)

Reply:
thinking especially re: the image list and image delete scenario

Reply:
Yeah, so originally this was supposed to actually only return the direct children (at least this is what the graphdriver implementation does). The usage in `daemon/list.go` only expects the direct children (but it recurses anyway so it's fine either way).

The image delete scenario needs to be checked though.

Reply:
Delete also calls `Children` and doesn't need ancestors, only direct children